File: File-Handling/remove_duplicates.py
import os
import time
from multiprocessing import Pool

def remove_duplicates(input_file):
    start_time = time.time()  # Record the start time
    output_folder = "C:/updates"
    output_file = os.path.join(output_folder, os.path.splitext(os.path.basename(input_file))[0] + "_unique.txt")
    total_records = 0
    total_unique_records = 0
    total_duplicate_records = 0
    seen = set()
    with open(input_file, 'r') as infile, open(output_file, 'w') as outfile:
        for line in infile:
            total_records += 1  # Increment total records count
            if line not in seen:
                seen.add(line)
                outfile.write(line)
                total_unique_records += 1  # Increment unique records count
            else:
                total_duplicate_records += 1  # Increment duplicate records count
    end_time = time.time()  # Record the end time
    execution_time = end_time - start_time  # Calculate the total execution time
    
    # Print counts
    print(f"Total records: {total_records}")
    print(f"Total unique records: {total_unique_records}")
    print(f"Total duplicate records: {total_duplicate_records}")
    print(f"Total execution time: {execution_time} seconds")
    print(f"Unique records file generated: {output_file}")

if __name__ == "__main__":
    input_files = ["C:/Logs/20millionloc.csv"]  # List of input files
   # input_files = ["C:/Logs/5m-records.csv"]  # List of input files
   # input_files = ["C:/Logs/500000-Sales-Records.csv"]  # List of input files
    num_processes = os.cpu_count()  # Number of CPU cores for parallel processing

    with Pool(num_processes) as pool:
        pool.map(remove_duplicates, input_files)

File: File-Handling/RemoveDuplicates.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.TimeUnit;
/**
*java -version
*javac RemoveDuplicates.java
*java RemoveDuplicates
*/

public class RemoveDuplicates {

    public static void main(String[] args) {
        String[] inputFiles = { "C:/Logs/20millionloc.csv" }; // List of input files
        // String[] inputFiles = { "C:/Logs/5m-records.csv" }; // List of input files
        // String[] inputFiles = { "C:/Logs/500000-Sales-Records.csv" }; // List of input files
        int numThreads = Runtime.getRuntime().availableProcessors(); // Number of CPU cores for parallel processing

        ExecutorService executor = Executors.newFixedThreadPool(numThreads);
        for (String inputFile : inputFiles) {
            executor.submit(new RemoveDuplicatesTask(inputFile));
        }
        executor.shutdown();
        try {
            executor.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    static class RemoveDuplicatesTask implements Callable<Void> {
        private final String inputFile;

        public RemoveDuplicatesTask(String inputFile) {
            this.inputFile = inputFile;
        }

        @Override
        public Void call() throws Exception {
            long startTime = System.currentTimeMillis();
            String outputFolder = "C:/updates";
            Path outputPath = Paths.get(outputFolder, Paths.get(inputFile).getFileName().toString().replaceFirst("[.][^.]+$", "_unique.txt"));
            int totalRecords = 0;
            int totalUniqueRecords = 0;
            int totalDuplicateRecords = 0;
            Set<String> seen = new HashSet<>();

            try (BufferedReader reader = new BufferedReader(new FileReader(inputFile));
                    BufferedWriter writer = new BufferedWriter(new FileWriter(outputPath.toString()))) {
                String line;
                while ((line = reader.readLine()) != null) {
                    totalRecords++;
                    if (!seen.contains(line)) {
                        seen.add(line);
                        writer.write(line);
                        writer.newLine();
                        totalUniqueRecords++;
                    } else {
                        totalDuplicateRecords++;
                    }
                }
            } catch (IOException e) {
                e.printStackTrace();
            }

            long endTime = System.currentTimeMillis();
            double executionTime = (endTime - startTime) / 1000.0;

            // Print counts
            System.out.println("Total records: " + totalRecords);
            System.out.println("Total unique records: " + totalUniqueRecords);
            System.out.println("Total duplicate records: " + totalDuplicateRecords);
            System.out.println("Total execution time: " + executionTime + " seconds");
            System.out.println("Unique records file generated: " + outputPath);

            return null;
        }
    }
}
